protocol/notify/src/main/kotlin/com/walletconnect/notify/engine/calls/GetNotificationHistoryUseCase.kt

```
@@ -0,0 +1,122 @@
+        val result = MutableStateFlow<GetNotificationHistory>(GetNotificationHistory.Processing)
+        var timeoutInfo: TimeoutInfo = TimeoutInfo.Nothing
+        try {
+            logger.log("getNotificationHistory - topic: $topic, limit: $limit, startingAfter: $startingAfter, timeout: $timeout")
+            val validTimeout = timeout.validateTimeout()
+            val subscription = subscriptionRepository.getActiveSubscriptionByNotifyTopic(topic)
+                ?: throw IllegalStateException("No subscription found for topic $topic")
+            val metadata: AppMetaData = metadataStorageRepository.getByTopicAndType(subscription.topic, AppMetaDataType.PEER)
+                ?: throw IllegalStateException("No metadata found for topic $topic")
+
+            val parsedLimit = minOf(MAX_LIMIT, limit ?: DEFAULT_LIMIT)
+
+            val sortedStoredNotifications = notificationsRepository.getNotificationsByTopic(topic).map { it.copy(metadata = metadata) }
+
+            try {
+                if (sortedStoredNotifications.size >= parsedLimit) {
+                    val indexOfAfter = startingAfter?.let { sortedStoredNotifications.indexOfFirst { it.id == startingAfter } } ?: 0
+                    val pickedNotifications = sortedStoredNotifications.subList(indexOfAfter, indexOfAfter + parsedLimit)
+
+                    if (pickedNotifications.size <= parsedLimit) {
+                        val hasMore = ((sortedStoredNotifications.size > pickedNotifications.size) || (pickedNotifications.size == parsedLimit)) && !pickedNotifications.last().isLast
+                        logger.log("getNotificationHistory local - size: $parsedLimit, hasMore: $hasMore")
+                        return@supervisorScope GetNotificationHistory.Success(pickedNotifications, hasMore)
+                    }
+                }
+            } catch (_: IndexOutOfBoundsException) {
+            }
```

Empty catch. We should handle this error. If this is called by a user we should return an error

---

Adding todo comment. Will change the logic here in new PR before the release.

---

Added